File: SEED 2.0/Network Security Labs/RSA Public-Key Encryption and Signature Lab 2.0/task4.cpp
#include <iostream>
#include <openssl/bn.h>
using namespace std;

void signMessage() {
    BIGNUM *d = BN_new();

    // Hexadecimal value of d (private key)
    BN_hex2bn(&d, "74D806F9F3A62BAE331FFE3F0A68AFE35B3D2E4794148AACBC26AA381CD7D30D");

    // Original message
    string M_original_hex = "49206f776520796f752024323030302e";  // Hexadecimal of "I owe you $2000."
    BIGNUM *M_original_hash = BN_new();
    BN_hex2bn(&M_original_hash, M_original_hex.c_str());

    // Sign the original message (S_original = M^d mod n)
    BIGNUM *S_original = BN_new();
    BN_mod_exp(S_original, M_original_hash, d, NULL, NULL);

    // Print the original signature
    char *S_original_hex = BN_bn2hex(S_original);
    cout << "Signature for the original message: " << S_original_hex << endl;

    // Clean up
    OPENSSL_free(S_original_hex);
    BN_free(d);
    BN_free(M_original_hash);
    BN_free(S_original);
}

int main() {
    signMessage();
    return 0;
}

File: SEED 2.0/Network Security Labs/RSA Public-Key Encryption and Signature Lab 2.0/task4.py
# Original message
M_original = "I owe you $2000."

# Provided private key (d) and public key (n)
d_hex = "74D806F9F3A62BAE331FFE3F0A68AFE35B3D2E4794148AACBC26AA381CD7D30D"
n_hex = "DCBFFE3E51F62E09CE7032E2677A78946A849DC4CDDE3A4D0CB81629242FB1A5"

d = int(d_hex, 16)
n = int(n_hex, 16)

# Sign the original message
M_original_hash = int(M_original.encode('utf-8').hex(), 16)
S_original = pow(M_original_hash, d, n)

# Print the original signature
print("Signature for the original message:", hex(S_original)[2:])

# Modify the message
M_modified = "I owe you $3000."

# Sign the modified message
M_modified_hash = int(M_modified.encode('utf-8').hex(), 16)
S_modified = pow(M_modified_hash, d, n)

# Print the modified signature
print("Signature for the modified message:", hex(S_modified)[2:])
